README.md
Scriptlike
==========

Utility to aid in writing script-like programs in the [D Programming Language](http://dlang.org).

This library has no external dependencies.

Tested with DMD 2.066.0 through 2.068.0.

* [API Reference](http://semitwist.com/scriptlike)
* [Changelog](http://semitwist.com/scriptlike/changelog.html)
* [DUB](http://code.dlang.org/about) [Package](http://code.dlang.org/packages/scriptlike)
* [Small article explaining the original motivations behind scriptlike](http://semitwist.com/articles/article/view/scriptlike-shell-scripting-in-d-annoyances-and-a-library-solution)

Importing
---------
```import scriptlike;```

That imports all of Scriptlike, plus automatically includes anything from Phobos likely to be useful for scripts.

Or, if you don't want any of Phobos imported automatically, you can import only Scriptlike:

```import scriptlike.only;```

Features
--------
* A thin wrapper over [std.path](http://dlang.org/phobos/std_path.html) and [std.file](http://dlang.org/phobos/std_file.html) that provides a dedicated Path type specifically designed for managing file paths in a simple, reliable, cross-platform way. No more dealing with slashes, paths-with-spaces, calling [buildPath](http://dlang.org/phobos/std_path.html#buildPath), [normalizing](http://dlang.org/phobos/std_path.html#buildNormalizedPath), or getting paths mixed up with ordinary strings.
* Easy user-input prompts with [interact](http://semitwist.com/scriptlike/scriptlike/interact.html) module.
* Optionally enable automatic command echoing (including shell commands, changing/creating directories and deleting/copying/moving/linking/renaming both directories and files) by setting one simple flag: [bool scriptlikeEcho](http://semitwist.com/scriptlike/scriptlike/core/scriptlikeEcho.html)
* Most typical Phobos modules automatically imported [unless you don't want to](http://semitwist.com/scriptlike/scriptlike/only.html). Who needs rows and rows of standard lib imports for a mere script?
* Less-pedantic filesystem operations for when you don't care whether it exists or not: [existsAsFile](http://semitwist.com/scriptlike/scriptlike/file/extras/existsAsFile.html), [existsAsDir](http://semitwist.com/scriptlike/scriptlike/file/extras/existsAsDir.html), [existsAsSymlink](http://semitwist.com/scriptlike/scriptlike/file/extras/existsAsSymlink.html), [tryRename](http://semitwist.com/scriptlike/scriptlike/file/extras/tryRename.html), [trySymlink](http://semitwist.com/scriptlike/scriptlike/file/extras/trySymlink.html), [tryCopy](http://semitwist.com/scriptlike/scriptlike/file/extras/tryCopy.html), [tryMkdir](http://semitwist.com/scriptlike/scriptlike/file/extras/tryMkdir.html), [tryMkdirRecurse](http://semitwist.com/scriptlike/scriptlike/file/extras/tryMkdirRecurse.html), [tryRmdir](http://semitwist.com/scriptlike/scriptlike/file/extras/tryRmdir.html), [tryRmdirRecurse](http://semitwist.com/scriptlike/scriptlike/file/extras/tryRmdirRecurse.html), [tryRemove](http://semitwist.com/scriptlike/scriptlike/file/extras/tryRemove.html): All check whether the source path exists and return WITHOUT throwing if there's nothing to do.
* One simple call, [run](http://semitwist.com/scriptlike/scriptlike/process/run.html), to run a shell command script-style (ie, synchronously with forwarded stdout/in/err) from any working directory. Or [runCollect](http://semitwist.com/scriptlike/scriptlike/process/runCollect.html) to capture the output instead of displaying it.
* [run](http://semitwist.com/scriptlike/scriptlike/process/run.html) and [runCollect](http://semitwist.com/scriptlike/scriptlike/process/runCollect.html) automatically throw if a process exits with a non-zero exit code. Or use [tryRun](http://semitwist.com/scriptlike/scriptlike/process/tryRun.html) and [tryRunCollect](http://semitwist.com/scriptlike/scriptlike/process/tryRunCollect.html) to ignore or handle the error level yourself.
* One simple function, [fail(...)](http://semitwist.com/scriptlike/scriptlike/fail/fail.html), to help you exit with an error message in an exception-safe way. Doesn't require *any* boilerplate in your main()!. Or use it std.exception.enforce-style with [failEnforce(cond, ...)](http://semitwist.com/scriptlike/scriptlike/fail/failEnforce.html).
* [More to come!](https://github.com/Abscissa/scriptlike/issues)


CHANGELOG.md
Scriptlike - Changelog
======================

(Dates below are YYYY/MM/DD)

v0.8.2 - TBD
-------------------
- **Enhancement:** Add (opt-in) command echoing to most functions in [```scriptlike.file```](http://semitwist.com/scriptlike/scriptlike/file.html).
- **Enhancement:** Add [```yap```](http://semitwist.com/scriptlike/scriptlike/file.html#yap) and [```yapFunc```](http://semitwist.com/scriptlike/scriptlike/file.html#yapFunc) as improved versions of to-be-deprecated [```echoCommand```](http://semitwist.com/scriptlike/scriptlike/file.html#echoCommand).
- **Fixed:** Make [```escapeShellArg```](http://semitwist.com/scriptlike/scriptlike/path.html#escapeShellArg) const-correct.

v0.8.1 - 2015/06/22
-------------------
- **Enhancement:** New overload for [```userInput```](http://semitwist.com/scriptlike-docs/v0.8.1/scriptlike/interact/userInput.html) to allow type inference: ```void userInput(T=string)(string question, ref T result);``` (suggestion from [Per Nordlöw](http://forum.dlang.org/post/povoxkcogcmbvhwlxqbc@forum.dlang.org)).

v0.8.0 - 2015/06/13
-------------------
- **Change:** Minimum officially supported DMD increased from v2.064.2 to v2.066.0. Versions below v2.066.0 may still work, but there will now be certain problems when dealing with paths that contain spaces, particularly on Windows.
- **Change:** Removed unnecessary non-[```Path```](http://semitwist.com/scriptlike/scriptlike/path.html#Path) wrappers around [```std.file```](http://dlang.org/phobos/std_file.html)/[```std.path```](http://dlang.org/phobos/std_path.html). Things not wrapped (like [```dirSeparator```](http://dlang.org/phobos/std_path.html#dirSeparator) and [```SpanMode```](http://dlang.org/phobos/std_file.html#SpanMode)) are now selective public imports instead of aliases. These changes should reduce issues with symbol conflicts.
- **Change:** [API reference](http://semitwist.com/scriptlike/) now built using [ddox](https://github.com/rejectedsoftware/ddox) and uses much improved styling (actually uses a stylesheet now).
- **Change:** Eliminate remnants of the "planned but never enabled" wstring/dstring versions of [```Path```](http://semitwist.com/scriptlike/scriptlike/path.html#Path)/[```Ext```](http://semitwist.com/scriptlike/scriptlike/path.html#Ext)/[```Args```](http://semitwist.com/scriptlike/scriptlike/process.html#Args). There turned out not to be much need for them, and even [```std.file```](http://dlang.org/phobos/std_file.html) doesn't support wstring/dstring either.
- **Change:** Put output binaries in "bin" subdirectory, instead of Scriptlike's root.
- **Enhancement:** Add module [```scriptlike.only```](http://semitwist.com/scriptlike/scriptlike/only.html) to import all of scriptlike, but omit the helper Phobos imports in [```scriptlike.std```](http://semitwist.com/scriptlike/scriptlike/std.html).
- **Enhancement:** [```fail```](http://semitwist.com/scriptlike/scriptlike/fail.html#fail) now accepts an arbitrary list of args of any type, just like writeln,
- **Enhancement:** Added [```failEnforce```](http://semitwist.com/scriptlike/scriptlike/fail.html#failEnforce), like std.exception.enforce, but for [```fail```](http://semitwist.com/scriptlike/scriptlike/fail.html#fail).
- **Enhancement:** Added [```runCollect```](http://semitwist.com/scriptlike/scriptlike/process.html#runCollect) and [```tryRunCollect```](http://semitwist.com/scriptlike/scriptlike/process.html#tryRunCollect), to capture a command's output instead of displaying it.
- **Enhancement:** Added [```pause```](http://semitwist.com/scriptlike/scriptlike/interact.html#pause) to pause and prompt the user to press Enter.
- **Enhancement:** [```echoCommand```](http://semitwist.com/scriptlike/scriptlike/file.html#echoCommand) is no longer private.
- **Enhancement:** Added [```Path```](http://semitwist.com/scriptlike/scriptlike/path.html#Path)-based wrappers for [```std.file```](http://dlang.org/phobos/std_file.html)'s [```getcwd```](http://semitwist.com/scriptlike/scriptlike/file.html#getcwd), [```thisExePath```](http://semitwist.com/scriptlike/scriptlike/file.html#thisExePath) and [```tempDir```](http://semitwist.com/scriptlike/scriptlike/file.html#tempDir).
- **Fixed:** No longer uses the deprecated [```std.process.system()```](http://dlang.org/phobos/std_process.html#system).

v0.7.0 - 2015/04/02
-------------------
- **Enhancement:** [#14](https://github.com/Abscissa/scriptlike/issues/14): Added [```interact```](http://semitwist.com/scriptlike/scriptlike/interact.html) module for easy user-input prompts. (Thanks to [Jesse Phillips](https://github.com/JesseKPhillips))
- **Fixed:** Unittest compile failure on DMD v2.067.0.

v0.6.0 - 2014/02/16
-------------------
- **Change:** [```Path```](http://semitwist.com/scriptlike/scriptlike/path.html#Path) and [```Ext```](http://semitwist.com/scriptlike/scriptlike/path.html#Ext) are now aliases for the UTF-8 instantiations, and the template structs are now named ```PathT``` and ```ExtT```.
- **Change:** Removed ```path()``` and ```ext()``` helper functions to free up useful names from the namespace, since they are no longer needed. Use ```Path()``` and ```Ext()``` instead.
- **Change:** Internally split into separate modules, but uses package.d to preserve ```import scriptlike;```.
- **Change**: Rename ```escapeShellPath``` -> [```escapeShellArg```](http://semitwist.com/scriptlike/scriptlike/path.html#escapeShellArg).
- **Change**: Rename ```runShell``` -> [```tryRun```](http://semitwist.com/scriptlike/scriptlike/process.html#tryRun). Temporarily keep ```runShell``` as an alias.
- **Change**: Rename ```scriptlikeTraceCommands``` -> [```scriptlikeEcho```](http://semitwist.com/scriptlike/scriptlike/file.html#scriptlikeEcho). Temporarily keep ```scriptlikeTraceCommands``` as an alias.
- **Enhancement:** Added scripts to run unittests and build API docs.

ddoc/changelog.d
/++
Scriptlike_Changelog:

The latest version of this changelog is always available at:$(BR)
$(LINK http://semitwist.com/scriptlike/changelog.html )

(Dates below are YYYY/MM/DD)

$(H2 v0.9.2 - 2015/07/01)

$(UL
	$(ENHANCE
		Add a "no-build" configuration for projects that need to import/depend
		on Scriptlike through DUB, but use their own buildsystem.
	)
)

$(H2 v0.9.1 - 2015/06/28)

$(UL
	$(FIXED Fails to compile unless the `makedocs` script has been run.)
)

$(H2 v0.9.0 - 2015/06/27)

$(UL
	$(CHANGE Split $(MODULE_FILE) and $(MODULE_PATH) into the following:$(BR)
		$(UL
			$(LI $(MODULE_CORE) )
			$(LI $(MODULE_FILE_EXTR) )
			$(LI $(MODULE_FILE_WRAP) )
			$(LI $(MODULE_PATH_EXTR) )
			$(LI $(MODULE_PATH_WRAP) )
		)
		Utilizes `package.d` to retain ability to import $(MODULE_FILE) and $(MODULE_PATH).
	)
	$(CHANGE Convert changelog from markdown to $(DDOX) so links are more readable. )
	$(ENHANCE Add (opt-in) command echoing to most functions in $(MODULE_FILE). )
	$(ENHANCE
		Add $(API_CORE yap) and $(API_CORE yapFunc) as improved versions
		of to-be-deprecated $(API_CORE echoCommand).
	)
	$(FIXED Make $(API_PATH_EXTR escapeShellArg) const-correct. )
	$(FIXED
		Make $(API_PATH_EXTR Path.toRawString) and $(API_PATH_EXTR Ext.toRawString)
		both be `pure @safe nothrow`.
	)
)

$(H2 v0.8.1 - 2015/06/22)

$(UL
	$(ENHANCE
		New overload for $(API_INTERACT userInput) to allow type inference:$(BR)
		`void userInput(T=string)(string question, ref T result);`
		(suggestion from
		$(LINK2 http://forum.dlang.org/post/povoxkcogcmbvhwlxqbc@forum.dlang.org, Per Nordlöw)).
	)
)

$(H2 v0.8.0 - 2015/06/13)

$(UL
	$(CHANGE
		Minimum officially supported $(DMD) increased from v2.064.2 to v2.066.0.
		Versions below v2.066.0 may still work, but there will now be certain
		problems when dealing with paths that contain spaces, particularly
		on Windows.
	)
	$(CHANGE
		Removed unnecessary non-$(API_PATH_EXTR Path) wrappers around $(MODULE_STD_FILE)/$(MODULE_STD_PATH).
		Things not wrapped (like $(STD_PATH dirSeparator) and $(STD_FILE SpanMode))
		are now selective public imports instead of aliases. These changes should
		reduce issues with symbol conflicts.
	)
	$(CHANGE
		$(LINK2 http://semitwist.com/scriptlike/, API reference) now built
		using $(DDOX) and uses much improved styling (actually uses a stylesheet now).
	)
	$(CHANGE
		Eliminate remnants of the "planned but never enabled" wstring/dstring
		versions of $(API_PATH_EXTR Path)/$(API_PATH_EXTR Ext)/$(API_PROCESS Args). There
		turned out not to be much need for them, and even $(MODULE_STD_FILE)
		doesn't support wstring/dstring either.
	)
	$(CHANGE Put output binaries in "bin" subdirectory, instead of Scriptlike's root. )
	$(ENHANCE
		Add module scriptlike.only to import all of scriptlike, but omit the
		helper Phobos imports in scriptlike.std.
	)
	$(ENHANCE
		$(API_FAIL fail) now accepts an arbitrary list of args of any type,
		just like $(STD_STDIO writeln),
	)
	$(ENHANCE
		Added $(API_FAIL failEnforce), like Phobos's $(STD_EXCEPTION enforce),
		but for $(API_FAIL fail).
	)
	$(ENHANCE
		Added $(API_PROCESS runCollect) and $(API_PROCESS tryRunCollect), to
		capture a command's output instead of displaying it.
	)
	$(ENHANCE Added $(API_INTERACT pause) to pause and prompt the user to press Enter. )
	$(ENHANCE $(API_CORE echoCommand) is no longer private. )
	$(ENHANCE
		Added $(API_PATH_EXTR Path)-based wrappers for $(MODULE_STD_FILE)'s 
		$(STD_FILE getcwd), $(STD_FILE thisExePath) and $(STD_FILE tempDir).
	)
	$(FIXED No longer uses Phobos's deprecated $(STD_PROCESS system) function.)
)

$(H2 v0.7.0 - 2015/04/02)

$(UL
	$(ENHANCE
		$(ISSUE 14): Added scriptlike.interact module for easy user-input prompts.
		(Thanks to $(LINK2 https://github.com/JesseKPhillips, Jesse Phillips))
	)
	$(FIXED Unittest compile failure on $(DMD) v2.067.0. )
)

$(H2 v0.6.0 - 2014/02/16)

$(UL
	$(CHANGE
		$(API_PATH_EXTR Path) and $(API_PATH_EXTR Ext) are now aliases for the UTF-8
		instantiations, and the template structs are now named `PathT` and `ExtT`.
	)
	$(CHANGE
		Removed `path()` and `ext()` helper functions to free up useful names
		from the namespace, since they are no longer needed. Use `Path()` and
		`Ext()` instead.
	)
	$(CHANGE
		Internally split into separate modules, but uses `package.d` to
		preserve `import scriptlike;`.
	)
	$(CHANGE Rename `escapeShellPath` -> $(API_PATH_EXTR escapeShellArg). )
	$(CHANGE
		Rename $(API_PROCESS runShell) -> $(API_PROCESS tryRun). Temporarily keep
		$(API_PROCESS runShell) as an alias.
	)
	$(CHANGE
		Rename $(API_CORE scriptlikeTraceCommands) -> $(API_CORE scriptlikeEcho).
		Temporarily keep $(API_CORE scriptlikeTraceCommands) as an alias.
	)
	$(ENHANCE Added scripts to run unittests and build API docs. )
	$(ENHANCE
		Added $(API_PATH_EXTR Path.opCast) and $(API_PATH_EXTR Ext.opCast) for
		converting to bool.
	)
	$(ENHANCE
		$(API_FAIL fail) no longer requires any boilerplate in `main()`.
		($(LINK2 http://forum.dlang.org/thread/ldc6qt$(DOLLAR)22tv$(DOLLAR)1@digitalmars.com, Newsgroup link))
	)
	$(ENHANCE
		Added $(API_PROCESS run) to run a shell command like $(API_PROCESS tryRun),
		but automatically throw if the process returns a non-zero error level.
	)
	$(ENHANCE $(ISSUE 2): Optional callback sink for command echoing: $(API_CORE scriptlikeCustomEcho). )
	$(ENHANCE $(ISSUE 8): Dry run support via bool $(API_CORE scriptlikeDryRun). )
	$(ENHANCE
		$(ISSUE 13): Added `ArgsT` (and $(API_PROCESS Args) helper alias)
		to safely build command strings from parts.
	)
	$(ENHANCE Added this changelog. )
	$(FIXED
		$(API_PATH_EXTR Path)(null) and $(API_PATH_EXTR Ext)(null) were automatically
		changed to empty string.
	)
	$(FIXED $(ISSUE 10): Docs should include all OS-specific functions. )
)

$(H2 v0.5.0 - 2014/02/11)

$(UL
	$(LI Initial release. )
)

Copyright:
Copyright (C) 2014-2015 Nick Sabalausky.
Portions Copyright (C) 2010 Jesse Phillips.

License: zlib/libpng
Authors: Nick Sabalausky, Jesse Phillips
+/
module changelog;
